File: packages/go/health.go
package rdk

import (
	"context"
	"fmt"
	"time"
)

// RollupHealth is a consolidated, beginner-friendly read of a rollup's status,
// its latest settlement batch, and (for optimistic rollups) the challenge-window
// countdown.
type RollupHealth struct {
	RollupID string
	// Status is the rollup lifecycle status (active/paused/stopped/pending).
	Status string
	// HasBatches reports whether any batch has been submitted.
	HasBatches bool
	// LatestBatchIndex is the latest batch index (valid only when HasBatches).
	LatestBatchIndex int
	// LatestBatchStatus is the latest batch status (valid only when HasBatches).
	LatestBatchStatus string
	// BatchAgeSecs is seconds since the latest batch was submitted.
	BatchAgeSecs int64
	// ChallengeDeadlineSecs is when the challenge window closes (unix seconds).
	ChallengeDeadlineSecs int64
	// SecondsUntilChallengeDeadline is seconds until the window closes (negative
	// if past).
	SecondsUntilChallengeDeadline int64
	// HasChallengeWindow reports whether the challenge-window fields are set.
	HasChallengeWindow bool
	// Healthy is a coarse health flag: active rollup, latest batch not rejected.
	Healthy bool
	// Notes are human-readable observations.
	Notes []string
}

// HealthOptions configures GetRollupHealth.
type HealthOptions struct {
	// NowSecs overrides the clock (unix seconds). Zero uses time.Now.
	NowSecs int64
}

// GetRollupHealth assembles a RollupHealth snapshot for a rollup.
func GetRollupHealth(ctx context.Context, client *RdkClient, rollupID string, options HealthOptions) (RollupHealth, error) {
	nowSecs := options.NowSecs
	if nowSecs == 0 {
		nowSecs = time.Now().Unix()
	}
	notes := []string{}

	rollup, err := client.Rest.GetRollup(ctx, rollupID)
	if err != nil {
		return RollupHealth{}, err
	}
	healthy := rollup.Status == "active"
	if rollup.Status != "active" {
		notes = append(notes, fmt.Sprintf("rollup status is %q", rollup.Status))
	}

	health := RollupHealth{RollupID: rollupID, Status: rollup.Status, HasBatches: false, Healthy: healthy, Notes: notes}

	latest, err := client.Rest.GetLatestBatch(ctx, rollupID)
	if err != nil || latest.SubmittedAt == 0 {
		health.Notes = append(health.Notes, "no settlement batches submitted yet")
		return health, nil
	}

	health.HasBatches = true
	health.LatestBatchIndex = latest.BatchIndex
	health.LatestBatchStatus = latest.Status
	health.BatchAgeSecs = nowSecs - int64(latest.SubmittedAt)

	if latest.Status == "rejected" {
		healthy = false
		health.Notes = append(health.Notes, "latest batch was rejected")
	}

	if latest.Status == "submitted" || latest.Status == "challenged" {
		params, perr := client.Params(ctx)
		if perr == nil {
			deadline := int64(latest.SubmittedAt) + int64(params.DefaultChallengeWindow)
			health.ChallengeDeadlineSecs = deadline
			health.SecondsUntilChallengeDeadline = deadline - nowSecs
			health.HasChallengeWindow = true
			if latest.Status == "challenged" {
				health.Notes = append(health.Notes, "latest batch is under challenge")
			}
		}
	}

	health.Healthy = healthy
	return health, nil
}

File: packages/rust/src/health.rs
//! Rollup health: a consolidated, beginner-friendly read of a rollup's status,
//! its latest settlement batch, and (for optimistic rollups) the
//! challenge-window countdown -- assembled from the existing read surface.

use crate::client::rest::RestError;
use crate::client::RdkClient;

/// A consolidated rollup-health snapshot.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RollupHealth {
    /// The rollup id.
    pub rollup_id: String,
    /// Rollup lifecycle status (active/paused/stopped/pending).
    pub status: String,
    /// Whether any batch has been submitted.
    pub has_batches: bool,
    /// The latest batch index, if any.
    pub latest_batch_index: Option<i64>,
    /// The latest batch status, if any.
    pub latest_batch_status: Option<String>,
    /// Seconds since the latest batch was submitted.
    pub batch_age_secs: Option<i64>,
    /// For optimistic batches: when the challenge window closes (unix seconds).
    pub challenge_deadline_secs: Option<i64>,
    /// Seconds until the challenge window closes (negative if already past).
    pub seconds_until_challenge_deadline: Option<i64>,
    /// Coarse health flag: active rollup, latest batch not rejected.
    pub healthy: bool,
    /// Human-readable observations.
    pub notes: Vec<String>,
}

/// Assemble a [`RollupHealth`] snapshot for a rollup.
pub fn get_rollup_health(
    client: &RdkClient,
    rollup_id: &str,
    now_secs: i64,
) -> Result<RollupHealth, RestError> {
    let mut notes: Vec<String> = Vec::new();
    let rollup = client.rest.get_rollup(rollup_id)?;
    let mut healthy = rollup.status == "active";
    if rollup.status != "active" {
        notes.push(format!("rollup status is \"{}\"", rollup.status));
    }

    let mut health = RollupHealth {
        rollup_id: rollup_id.to_string(),
        status: rollup.status.clone(),
        has_batches: false,
        latest_batch_index: None,
        latest_batch_status: None,
        batch_age_secs: None,
        challenge_deadline_secs: None,
        seconds_until_challenge_deadline: None,
        healthy,
        notes,
    };

    let latest_batch = match client.rest.get_latest_batch(rollup_id) {
        Ok(b) if b.submitted_at != 0 => b,
        _ => {
            health
                .notes
                .push("no settlement batches submitted yet".to_string());
            return Ok(health);
        }
    };

    health.has_batches = true;
    health.latest_batch_index = Some(latest_batch.batch_index);
    health.latest_batch_status = Some(latest_batch.status.clone());
    health.batch_age_secs = Some(now_secs - latest_batch.submitted_at);

    if latest_batch.status == "rejected" {
        healthy = false;
        health.notes.push("latest batch was rejected".to_string());
    }

    if latest_batch.status == "submitted" || latest_batch.status == "challenged" {
        let params = client.params()?;
        let deadline = latest_batch.submitted_at + params.default_challenge_window;
        health.challenge_deadline_secs = Some(deadline);
        health.seconds_until_challenge_deadline = Some(deadline - now_secs);
        if latest_batch.status == "challenged" {
            health
                .notes
                .push("latest batch is under challenge".to_string());
        }
    }

    health.healthy = healthy;
    Ok(health)
}
